src/log/Entry.h

```diff
@@ -4,103 +4,112 @@
 #ifndef __CEPH_LOG_ENTRY_H
 #define __CEPH_LOG_ENTRY_H
 
-#include "common/CachedPrebufferedStreambuf.h"
-#include <pthread.h>
-#include <string>
 #include "log/LogClock.h"
 
+#include "common/StackStringStream.h"
 
-namespace ceph {
-namespace logging {
+#include "boost/container/small_vector.hpp"
 
-struct Entry {
-  log_time m_stamp;
-  pthread_t m_thread;
-  short m_prio, m_subsys;
-  Entry *m_next;
+#include <pthread.h>
 
-  size_t m_buf_len;
-  size_t* m_exp_len;
-  char m_static_buf[1];
+#include <string_view>
 
-  prebuffered_data m_data;
-  CachedPrebufferedStreambuf* m_streambuf;
+namespace ceph {
+namespace logging {
 
-  Entry()
-    : Entry(log_time{}, 0, 0, 0, nullptr)
-  {}
-  Entry(log_time s, pthread_t t, short pr, short sub,
-	const char *msg = NULL)
-    : Entry(s, t, pr, sub, m_static_buf, sizeof(m_static_buf), nullptr,
-	    msg)
+class Entry {
+public:
+  using time = log_time;
+
+  Entry() = delete;
+  Entry(short pr, short sub) :
+    m_stamp(clock().now()),
+    m_thread(pthread_self()),
+    m_prio(pr),
+    m_subsys(sub)
   {}
-  Entry(log_time s, pthread_t t, short pr, short sub, char* buf, size_t buf_len, size_t* exp_len,
-	const char *msg = NULL)
-    : m_stamp(s), m_thread(t), m_prio(pr), m_subsys(sub),
-      m_next(NULL),
-      m_buf_len(buf_len),
-      m_exp_len(exp_len),
-      m_data(buf, buf_len),
-      m_streambuf(CachedPrebufferedStreambuf::create(&m_data))
-  {
-    if (msg) {
-      get_ostream() << msg;
-    }
-  }
+  Entry(const Entry &) = default;
+  Entry& operator=(const Entry &) = default;
+  Entry(Entry &&e) = default;
+  Entry& operator=(Entry &&e) = default;
+  virtual ~Entry() = default;
+
+  virtual std::string_view strv() const = 0;
+  virtual std::size_t size() const = 0;
```

Comment:
as @tchaikov mentioned in https://github.com/ceph/ceph/pull/23721#discussion_r213147412, there is still some overhead from these virtual functions. i think we could remove its vtable entirely if `Log::submit_entry()` took `MutableEntry&&` directly, and `ConcreteEntry`'s converting constructor took `const MutableEntry&` instead of `const Entry&`. think that's worth doing?

Reply:
I just templatized `submit_entry`. Does it look okay now?

Reply:
it looks like `m_new.emplace_back(std::move(e))` still calls `ConcreteEntry(const Entry&)`, which relies on the virtual `Entry::strv()` to perform the copy. if it was `ConcreteEntry(const MutableEntry&)`, we wouldn't need any virtuals in `Entry`

Reply:
Jeeze, is this all really worth it? The cost to code readability seems really high.

Reply:
> think that's worth doing?

apparently not, and that's fine - i just wanted to clarify that the change to submit_entry didn't address @tchaikov's point about virtual function overhead

Reply:
Right, thanks for pointing it out! As you saw, I undid the templated submit_entry method. I don't think it's worth the trouble.